File: examples/nested-config/src/main/java/com/riversoforion/lena/example/ApplicationConfig.java
/*
 * Copyright (c) 2024-2025. Eric McIntyre / Rivers of Orion
 */
package com.riversoforion.lena.example;

import org.riversoforion.lena.config.ConfigurationProperties;
import org.riversoforion.lena.config.Name;
import org.riversoforion.lena.config.Namespace;

import static org.riversoforion.lena.config.ConfigurationSources.*;

public class ApplicationConfig extends ConfigurationProperties {

    private ServiceConfig service;
    private NetworkConfig net;

    public ApplicationConfig() {

        this(Namespace.root());
    }

    ApplicationConfig(Namespace testingNs) {

        super(prioritized(forEnvironment(), forSystemProperties()), testingNs);
        returnNullForMissing();
    }

    @Override
    protected void createChildren() {

        this.service = new ServiceConfig(namespace().child("service"));
        this.net = new NetworkConfig(namespace().child("net"));
    }

    public ServiceConfig service() {

        return service;
    }

    public NetworkConfig net() {

        return net;
    }

    public boolean isLocalMode() {

        return booleanVal(Name.of("local", "mode"));
    }
}

File: examples/nested-config/src/main/kotlin/com/riversoforion/lena/example/ApplicationConfig.kt
/*
 * Copyright (c) 2024-2026. Eric McIntyre / Rivers of Orion
 */
package com.riversoforion.lena.example

import org.riversoforion.lena.config.ConfigurationProperties
import org.riversoforion.lena.config.ConfigurationSource
import org.riversoforion.lena.config.ConfigurationSources.forEnvironment
import org.riversoforion.lena.config.ConfigurationSources.forSystemProperties
import org.riversoforion.lena.config.ConfigurationSources.prioritized
import org.riversoforion.lena.config.Namespace

/**
 * Exposes a set of environment variables and/or system properties with type-safe accessors,
 * including nested sub-configurations composed explicitly (no global registry).
 */
class ApplicationConfig(
    ns: Namespace = Namespace.root(),
    backingSource: ConfigurationSource = prioritized(forEnvironment(), forSystemProperties()),
) : ConfigurationProperties(backingSource, ns) {

    val service: ServiceConfig = ServiceConfig(backingSource, namespace.child("service"))
    val net: NetworkConfig = NetworkConfig(backingSource, namespace.child("net"))

    val localMode: Boolean by boolean("local", "mode")
}
